datalab/datalab_session/data_operations/data_operation.py

```diff
@@ -95,8 +94,13 @@ def set_output(self, output_data: dict):
 
     def get_output(self) -> dict:
         return cache.get(f'operation_{self.cache_key}_output')
+    
+    def set_failed(self, message: str):
+        self.set_status('FAILED')
+        self.set_message(message)
+        self.set_percent_completion(1.0)
```

Comment:
Shouldn't this be 0 since it didn't complete? Or maybe leave it in whatever percentage it was when it failed? Setting it to 1 seems weird though.

Reply:
I like the idea of leaving it at the percent where it failed. The front end deletes it soon after so it doesn't really matter but that makes sense.